spanconfig/spanconfigkvsubscriber: TestBlockedKVSubscriberDisablesQueues failed

spanconfig/spanconfigkvsubscriber.TestBlockedKVSubscriberDisablesQueues [failed](https://tanzanite.cluster.engflow.com/invocations/default/69e2fb92-3249-4b38-aa13-f0240e4620e1?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3BhbmNvbmZpZy9zcGFuY29uZmlna3ZzdWJzY3JpYmVyOnNwYW5jb25maWdrdnN1YnNjcmliZXJfdGVzdA==) on release-24.2.0-rc @ [3d1bcc1df630c10c4065a22fc846daa911b436d6](https://github.com/cockroachdb/cockroach/commits/3d1bcc1df630c10c4065a22fc846daa911b436d6):

Fatal error:

```
panic: test timed out after 41m37s
running tests:
	TestBlockedKVSubscriberDisablesQueues (40m41s)
	TestBlockedKVSubscriberDisablesQueues/queue=replicate (40m24s)

```
Stack: 

```
goroutine 26335 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x45
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/cockroach/pkg/sql/contention/contentionutils.(*ConcurrentBufferGuard).syncLocked(0xc00a3231f0)
* 	github.com/cockroachdb/cockroach/pkg/sql/contention/contentionutils/concurrent_buffer_guard.go:158 +0x52
* github.com/cockroachdb/cockroach/pkg/sql/contention/contentionutils.(*ConcurrentBufferGuard).ForceSync(0xc00a3231f0)
* 	github.com/cockroachdb/cockroach/pkg/sql/contention/contentionutils/concurrent_buffer_guard.go:123 +0x87
* github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights.(*concurrentBufferIngester).Start.func2({0xc004077c08?, 0x0?})
* 	github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights/pkg/sql/sqlstats/insights/ingester.go:120 +0x65
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x263
* created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 18997
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:476 +0x69d
* 
* goroutine 24360 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc00b67a750, {0xb416ac0, 0xc00b67eeb0})
* 	GOROOT/src/database/sql/sql.go:1246 +0xed
* created by database/sql.OpenDB in goroutine 18997
* 	GOROOT/src/database/sql/sql.go:824 +0x2d4
* 
* goroutine 26217 [sleep]:
* time.Sleep(0x4c4b40)
* 	GOROOT/src/runtime/time.go:195 +0x115
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Quiesce(0xc000e2f4a0, {0xb416af8, 0xc00d607230})
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:610 +0x176
* github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers.func1(0x1, 0xc000e2f4a0)
* 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:156 +0x17f
* created by github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers in goroutine 18997
* 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:152 +0x1cf
* 
* goroutine 26104 [select, 2 minutes]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:253 +0x19b
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 26031
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:248 +0xad
* 
* goroutine 26105 [sync.Cond.Wait, 2 minutes]:
* sync.runtime_notifyListWait(0xc00e2f40a8, 0x1)
* 	GOROOT/src/runtime/sema.go:569 +0x159
* sync.(*Cond).Wait(0xc00e2f4098)
* 	GOROOT/src/sync/cond.go:70 +0x75
* github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:369
* github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0xc00e2f4000, {0xc006b84140?, 0x7cef3a0?})
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:643 +0xb79
* runtime/pprof.Do({0xb416998, 0xfffae00}, {{0xc000cecfc0?, 0x20367a0?, 0xc005c4b4b0?}}, 0xc00c3aafc0)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:562 +0x92
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 26031
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:561 +0x8e5
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBlockedKVSubscriberDisablesQueues.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-41025

--- a/pkg/sql/sqlstats/insights/BUILD.bazel
+++ b/pkg/sql/sqlstats/insights/BUILD.bazel
@@ -50,13 +50,11 @@ go_test(
         "//pkg/settings/cluster",
         "//pkg/sql/appstatspb",
         "//pkg/sql/clusterunique",
-        "//pkg/testutils",
         "//pkg/util/leaktest",
         "//pkg/util/log",
         "//pkg/util/stop",
         "//pkg/util/uint128",
         "//pkg/util/uuid",
-        "@com_github_cockroachdb_errors//:errors",
         "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/sql/sqlstats/insights/ingester.go
+++ b/pkg/sql/sqlstats/insights/ingester.go
@@ -38,8 +38,9 @@ type ConcurrentBufferIngester struct {
 
 	eventBufferCh chan eventBufChPayload
 	registry      *lockingRegistry
-	running       uint64
 	clearRegistry uint32
+
+	closeCh chan struct{}
 }
 
 type eventBufChPayload struct {
@@ -89,7 +90,6 @@ func (i *ConcurrentBufferIngester) Start(
 	// This task pulls buffers from the channel and forwards them along to the
 	// underlying registry.
 	_ = stopper.RunAsyncTask(ctx, "insights-ingester", func(ctx context.Context) {
-		atomic.StoreUint64(&i.running, 1)
 
 		for {
 			select {
@@ -100,7 +100,7 @@ func (i *ConcurrentBufferIngester) Start(
 				}
 				eventBufferPool.Put(payload.events)
 			case <-stopper.ShouldQuiesce():
-				atomic.StoreUint64(&i.running, 0)
+				close(i.closeCh)
 				return
 			}
 		}
@@ -188,6 +188,7 @@ func newConcurrentBufferIngester(registry *lockingRegistry) *ConcurrentBufferIng
 		// adjusting our carrying capacity.
 		eventBufferCh: make(chan eventBufChPayload, 1),
 		registry:      registry,
+		closeCh:       make(chan struct{}),
 	}
 
 	i.guard.eventBuffer = eventBufferPool.Get().(*eventBuffer)
@@ -202,11 +203,12 @@ func newConcurrentBufferIngester(registry *lockingRegistry) *ConcurrentBufferIng
 					atomic.StoreUint32(&i.clearRegistry, 0)
 				}()
 			}
-			if atomic.LoadUint64(&i.running) == 1 {
-				i.eventBufferCh <- eventBufChPayload{
-					clearRegistry: clearRegistry,
-					events:        i.guard.eventBuffer,
-				}
+			select {
+			case i.eventBufferCh <- eventBufChPayload{
+				clearRegistry: clearRegistry,
+				events:        i.guard.eventBuffer,
+			}:
+			case <-i.closeCh:
 			}
 			i.guard.eventBuffer = eventBufferPool.Get().(*eventBuffer)
 		},

--- a/pkg/sql/sqlstats/insights/ingester_test.go
+++ b/pkg/sql/sqlstats/insights/ingester_test.go
@@ -12,19 +12,17 @@ package insights
 
 import (
 	"context"
-	"sync/atomic"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/clusterunique"
-	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/uint128"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
-	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
 
@@ -127,56 +125,53 @@ func TestIngester_Clear(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	settings := cluster.MakeTestingClusterSettings()
 
-	t.Run("clears buffer", func(t *testing.T) {
-		ctx := context.Background()
-		stopper := stop.NewStopper()
-		defer stopper.Stop(ctx)
-		store := newStore(settings)
-		ingester := newConcurrentBufferIngester(
-			newRegistry(settings, &fakeDetector{
-				stubEnabled: true,
-				stubIsSlow:  true,
-			}, store))
-		// Start the ingester and wait for its async task to start.
-		// Disable timed flushes, so we can guarantee the presence of data
-		// in the buffer later on.
-		ingester.Start(ctx, stopper, WithoutTimedFlush())
-		testutils.SucceedsSoon(t, func() error {
-			if !(atomic.LoadUint64(&ingester.running) == 1) {
-				return errors.New("ingester not yet started")
-			}
-			return nil
-		})
-		// Fill the ingester's buffer with some data. This sets us up to
-		// call Clear() with guaranteed data in the buffer, so we can assert
-		// afterward that it's been cleared.
-		ingesterObservations := []testEvent{
-			{sessionID: 1, statementID: 10},
-			{sessionID: 2, statementID: 20},
-			{sessionID: 1, statementID: 11},
-			{sessionID: 2, statementID: 21},
-			{sessionID: 1, transactionID: 100},
-			{sessionID: 2, transactionID: 200},
+	ctx := context.Background()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+
+	store := newStore(settings)
+	ingester := newConcurrentBufferIngester(
+		newRegistry(settings, &fakeDetector{
+			stubEnabled: true,
+			stubIsSlow:  true,
+		}, store))
+
+	// Fill the ingester's buffer with some data. This sets us up to
+	// call Clear() with guaranteed data in the buffer, so we can assert
+	// afterward that it's been cleared.
+	ingesterObservations := []testEvent{
+		{sessionID: 1, statementID: 10},
+		{sessionID: 2, statementID: 20},
+		{sessionID: 1, statementID: 11},
+		{sessionID: 2, statementID: 21},
+		{sessionID: 1, transactionID: 100},
+		{sessionID: 2, transactionID: 200},
+	}
+	for _, o := range ingesterObservations {
+		if o.statementID != 0 {
+			ingester.ObserveStatement(o.SessionID(), &Statement{ID: o.StatementID()})
+		} else {
+			ingester.ObserveTransaction(o.SessionID(), &Transaction{ID: o.TransactionID()})
 		}
-		for _, o := range ingesterObservations {
-			if o.statementID != 0 {
-				ingester.ObserveStatement(o.SessionID(), &Statement{ID: o.StatementID()})
-			} else {
-				ingester.ObserveTransaction(o.SessionID(), &Transaction{ID: o.TransactionID()})
-			}
-		}
-		empty := event{}
-		require.NotEqual(t, empty, ingester.guard.eventBuffer[0])
-		// Now, call Clear() to verify it clears the buffer.
-		// Use SucceedsSoon for assertions, since the operation is async.
-		ingester.Clear()
-		testutils.SucceedsSoon(t, func() error {
-			if ingester.guard.eventBuffer[0] != empty {
-				return errors.New("eventBuffer not empty")
-			}
-			return nil
-		})
-	})
+	}
+	empty := event{}
+	require.Empty(t, ingester.eventBufferCh)
+	require.NotEqual(t, empty, ingester.guard.eventBuffer[0])
+	// Now, call Clear() to verify it clears the buffer. This operation
+	// is synchronous here because `ingester.eventBufferCh` has a buffer
+	// of 1 so the `Clear` operation can write to it without requiring a
+	// corresponding insights ingester task running. We just check to
+	// make sure `Clear` results in something getting posted to the
+	// channel.
+	ingester.Clear()
+	require.Equal(t, empty, ingester.guard.eventBuffer[0])
+	require.NotEmpty(t, ingester.eventBufferCh)
+	recv := <-ingester.eventBufferCh
+	for i := range ingesterObservations {
+		require.NotEqual(t, empty, recv.events[i])
+	}
+	// events 0-5 contain the payloads above, rest are empty
+	require.Equal(t, empty, recv.events[6])
 }
 
 func TestIngester_Disabled(t *testing.T) {
@@ -193,6 +188,8 @@ func TestIngester_Disabled(t *testing.T) {
 }
 
 func TestIngester_DoesNotBlockWhenReceivingManyObservationsAfterShutdown(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
 	// We have seen some tests hanging in CI, implicating this ingester in
 	// their goroutine dumps. We reproduce what we think is happening here,
 	// observing high volumes of SQL traffic after our consumer has shut down.
@@ -245,4 +242,41 @@ func (s testEvent) TransactionID() uuid.UUID {
 
 func (s testEvent) StatementID() clusterunique.ID {
 	return clusterunique.ID{Uint128: uint128.FromInts(0, s.statementID)}
+}
+
+// We had an issue with the insights ingester flush task being blocked
+// forever on shutdown. This was because of a bug where the order of
+// operations during stopper quiescence could cause `ForceSync()` to be
+// triggered twice without an intervening ingest operation. The second
+// `ForceSync()` would block forever because the buffer channel has a
+// capacity of 1.
+func TestIngesterBlockedForceSync(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+
+	st := cluster.MakeTestingClusterSettings()
+	registry := newRegistry(st, &fakeDetector{stubEnabled: true}, newStore(st))
+	ingester := newConcurrentBufferIngester(registry)
+
+	// We queue up a bunch of sync operations because it's unclear how
+	// many will proceed between the `Start()` and `Stop()` calls below.
+	ingester.guard.ForceSync()
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < 3; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ingester.guard.ForceSync()
+		}()
+	}
+
+	ingester.Start(ctx, stopper, WithoutTimedFlush())
+	stopper.Stop(ctx)
+	<-stopper.IsStopped()
+	wg.Wait()
 }

